ENH Add periodic extrapolation to SplineTransformer

mlondschien opened the pull request:
#### Reference Issues/PRs
Extends #18368.

#### What does this implement/fix? Explain your changes.
This PR adds the option for periodic extrapolation (with periodic splines) to the `SplineTransformer`.

#### Any other comments?
Thank you @lorentzenchr for adding the `SplineTransformer`! I often use periodic splines e.g. to model effects for the day of the year and would believe this feature to be very helpful. Please let me know if we should expand documenting code / examples.

mlondschien committed:
Periodic splines.

mlondschien committed:
Add test.

mlondschien committed:
Flake8.

mlondschien committed:
Larger intercept, less knots in test.

mlondschien committed:
Do extrapolation by hand for scipy<1.6.1

mlondschien committed:
Remove breakpoint.

ogrisel commented:
If supporting old versions of scipy is too much of a hassle I think it's fine to raise `NotImplementedError` with a message that explains to upgrade scipy to use this new feature.

ogrisel commented:
Alternatively, if missing code in scipy is a rather self-contained pure-python function, feel free to backport it in `sklearn/utils/fixes.py` with a comment to credit provenance.

mlondschien commented:
> If supporting old versions of scipy is too much of a hassle I think it's fine to raise NotImplementedError with a message that explains to upgrade scipy to use this new feature.

> Alternatively, if missing code in scipy is a rather self-contained pure-python function, feel free to backport it in `sklearn/utils/fixes.py` with a comment to credit provenance.

The only code needed to support `scipy<1.6.1` is (basically a copy from the code in scipy):
```python
n = spl.t.size - spl.k - 1
X[:, i] = spl.t[spl.k] + (X[:, i] - spl.t[spl.k]) % (spl.t[n] - spl.t[spl.k])
```
IMO this is fine, but raising a `NotImplementedError` works as well. Adding a backport to `sklean/utils/fixes` requires more code than above and might be too much added complexity for my taste.

ogrisel commented:
Alright, I am fine with keeping this at it is then.

ogrisel commented:
Could you please extend the spline example with a new section on modeling periodic data (with synthetic data with a mix of yearly and weekly signals)?

https://scikit-learn.org/dev/auto_examples/linear_model/plot_polynomial_interpolation.html#sphx-glr-auto-examples-linear-model-plot-polynomial-interpolation-py

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -346,7 +364,18 @@ def transform(self, X):
+                    n = spl.t.size - spl.k - 1
+                    X[:, i] = spl.t[spl.k] + (X[:, i] - spl.t[spl.k]) % (
+                        spl.t[n] - spl.t[spl.k]
+                    )
```
Maybe you can add a test that checks that you can reproduce the same transformation with scipy >= 1.6.1 and `pytest.skip` otherwise.

mlondschien replied:
I'm not sure how to implement this. We could do some monkeypatching. Can you point me to such a test for a similar setting?

ogrisel replied:
Write a test that compares the out-of-range prediction of a `SplineTransformer(extrapolation="periodic")` transformer with the matching transformation computed using `BSpline(.., extrapolate="periodic")` from scipy on the same data (using a dataset with a single feature).

ogrisel replied:
Use only 1 or 2 knots to make the test as simple as possible.

mlondschien replied:
The minimal example is degree 2 with three knots. I've added a test.

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +168,37 @@ def test_spline_transformer_linear_regression(bias, intercept):
+@pytest.mark.parametrize(["bias", "intercept"], [(True, False), (False, True)])
+def test_spline_transformer_periodic_linear_regression(bias, intercept):
+    """Test that B-splines fit a periodic curve pretty well."""
+    # +2 to avoid the value 0 in assert_allclose
```
```suggestion
    # "+ 3" to avoid the value 0 in assert_allclose
```

mlondschien committed:
Update sklearn/preprocessing/tests/test_polynomial.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

mlondschien committed:
Avoid inplace operation, raise for too few knots.

mlondschien committed:
Test for ValueError.

mlondschien committed:
Merge branch 'splines-periodic' of github.com:mlondschien/scikit-learn into splines-periodic

mlondschien committed:
Flake8.

mlondschien committed:
Example.

mlondschien commented:
I've added the test and an example. I used a synthetic example as I believed that this better visualized how periodic splines work. I am open for other suggestions though. How do I check whether the example renders correctly?

mlondschien commented:
Since you implemented the transformer, it would be very helpful if you could have a look @lorentzenchr.

lorentzenchr commented:
> Since you implemented the transformer, it would be very helpful if you could have a look

@mlondschien I will. Thanks for working on this extension.

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -346,8 +368,23 @@ def transform(self, X):
-                XBS[:, (i * n_splines):((i + 1) * n_splines)] = spl(X[:, i])
+            if self.extrapolation in ("continue", "error", "periodic"):
+
+                if self.extrapolation == "periodic":
+                    # With periodic extrapolation we map x to the segment
+                    # [spl.t[k], spl.t[n]].
+                    # This is equivalent to BSpline(.., extrapolate="periodic")
+                    # for scipy>=1.6.1.
+                    n = spl.t.size - spl.k - 1
+                    # Assign to new array to avoid inplace operation
+                    x = spl.t[spl.k] + (X[:, i] - spl.t[spl.k]) % (
+                        spl.t[n] - spl.t[spl.k]
+                    )
```
Periodic BSplines were implemented in scipy 1.0.0 in https://github.com/scipy/scipy/pull/7185. If we were to increment our minimal scipy from 0.19 to 1.0 for our next release, scikit-learn 1.0, we would not need this code.

mlondschien replied:
The comment was incorrect. 1.6.1 is the current version. I've changed it.

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -75,7 +92,8 @@ def test_spline_transformer_input_validation(params, err_msg):
 
 
 def test_spline_transformer_manual_knot_input():
-    """Test that array-like knot positions in SplineTransformer are accepted.
+    """
+    Test that array-like knot positions in SplineTransformer are accepted.
```
Why?

mlondschien replied:
Flake8 was complaining. See https://github.com/scikit-learn/scikit-learn/pull/19483/commits/a6e79e2c9a1e9ae597a01e7163e044174576451c and https://github.com/scikit-learn/scikit-learn/pull/19483/commits/8d8a2195b09f1d2efeebe77178a32a1f51c1221a.

lorentzenchr reviewed (commented):
On the test side, I miss a test that the extrapolation works. Let's say period length is 1. Fitting is on x in [0, 1] then transform([0, 1]) == transform([1, 2]) and in particular at the knots.

Maybe, one should more clearly state, that the periodicity is given either by the manually specified knots or by the range of the training data. It is neither automatically detected, nor can you specify the periodicity directly (with a parameter).

mlondschien committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

mlondschien committed:
Implement suggestions by @lorentzenchr.

mlondschien commented on examples/linear_model/plot_polynomial_interpolation.py:
```diff
@@ -145,3 +145,60 @@ def f(x):
+# suited for extrapolation beyond the range of training observations. In some
+# settings, e.g. with seasonal effects, we expect periodic continuation of
+# the underlying signal. Such effects can be modelled using periodic splines,
+# which have equal intercepts and derivatives at the first and last knot.
```
Equal function value is better. I kept _derivatives_ vague as this depends on the degree. Do you have a suggestion to briefly and correctly specify this?

mlondschien commented on examples/linear_model/plot_polynomial_interpolation.py:
```diff
@@ -145,3 +145,60 @@ def f(x):
+ax.legend()
+fig.show()
+
+# %% We again plot the underlying splines.
+fig, ax = plt.subplots()
+knots = np.linspace(0, 2 * np.pi, 4)
+splt = SplineTransformer(
+  knots=knots[:, None],
+  degree=3,
+  extrapolation="periodic"
+).fit(X_train)
+ax.plot(x_plot, splt.transform(X_plot))
+ax.legend(ax.lines, [f"spline {n}" for n in range(3)])
+
+# plot knots of spline
+ax.vlines(knots, ymin=0, ymax=0.8, linestyles='dashed')
+plt.show()
```
We could plot the first and last knots (and their continuation) in a different color. I've removed it until now.

mlondschien commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -297,8 +315,12 @@ def fit(self, X, y=None):
-        extrapolate = self.extrapolation == "continue"
+        coef = np.eye(n_splines, dtype=np.float64)
+        if self.extrapolation == "periodic":
+            coef = np.concatenate((coef, coef[:degree, :]))
```
Flake8 was complaining. See https://github.com/scikit-learn/scikit-learn/pull/19483/commits/a6e79e2c9a1e9ae597a01e7163e044174576451c and https://github.com/scikit-learn/scikit-learn/pull/19483/commits/8d8a2195b09f1d2efeebe77178a32a1f51c1221a.

mlondschien replied:
Yes, very much so. The _i_-th coordinate that get's returned by `bsp(x)` is the sum of the _i_-th and _nsplines - degree + i_-th B-Spline, which guarantees the periodicity. Sadly I did not find any literature that explains this in a simple / concise manner.

mlondschien commented:
Thanks @lorentzenchr for having a look. 

> On the test side, I miss a test that the extrapolation works. Let's say period length is 1. Fitting is on x in [0, 1] then transform([0, 1]) == transform([1, 2]) and in particular at the knots.

This is done [in this test](https://github.com/scikit-learn/scikit-learn/pull/19483/files#diff-5527bc676d1f3e3d77d2964f863103039ef46f809db3a41ffaea58d5d686fb1bR217), even if not as explicitly as you are suggesting. However this (and the test you suggested) would pass for any periodic continuation of splines (e.g. normal B-Splines with `extrapolate="periodic"`. I can add a test that asserts that transformers with knots at `[1, 2, 3, 4, 5]` and `[2, 3, 4, 5, 6]` yield the same result up to permutation of coordinates.

> Maybe, one should more clearly state, that the periodicity is given either by the manually specified knots or by the range of the training data. It is neither automatically detected, nor can you specify the periodicity directly (with a parameter).

I have tried to make this more explicit in the example. Where / how would you specify this in the docstring? One could add

> If 'periodic', periodic splines with a periodicity equal to the distance of the first and last knot are used. Periodic splines enforce equal function values and derivatives at the first and last knots.

to the `extrapolation` parameter. Maybe this could be made more concise.

Note that [this statement](https://github.com/scikit-learn/scikit-learn/pull/19483/files#diff-690d1b546a74ed69a4dc3e77edaa88363fc0f8297a71b1a3619a0f1cf79b2621R25-R27) is no longer correct. Would you like to point out the different number of splines for periodic splines or is this a minor detail?

ogrisel committed:
Extend test range in example

ogrisel commented:
@mlondschien thanks for the example! I extended it a bit to better demonstrate the extra-polation properties.

Please edit the existing changelog entry about `SplineTransformer` in `doc/whats_new/v1.0.rst` to reference your PR number and append your name.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -250,19 +253,29 @@ def fit(self, X, y=None):
         n_knots = base_knots.shape[0]
+
+        if self.extrapolation == "periodic" and n_knots <= self.degree:
+            raise ValueError("Periodic splines require degree < n_knots.")
```
Please recall the effective values for `n_knots` and  and `degrees` in the error message.

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -1,10 +1,19 @@
 from sklearn.pipeline import Pipeline
 from sklearn.preprocessing import KBinsDiscretizer, SplineTransformer
+from sklearn.utils.fixes import sp_version
+
+try:
+    from pkg_resources import parse_version  # type: ignore
+except ImportError:
+    # setuptools not installed
+    parse_version = LooseVersion  # type: ignore
```
You can:

```suggestion
from sklearn.utils.fixes import parse_version
```

directly instead.

ogrisel reviewed (commented):
Would need more time to understand the maths but here is some feedback in the mean time.

mlondschien commented:
Thanks @ogrisel and @lorentzenchr for having another look. While implementing the test as suggested by @lorentzenchr I noticed that my implementation as is does not work for non-uniform knots. See the picture below:
![Figure_1](https://user-images.githubusercontent.com/61679398/109381854-0b711d00-78dd-11eb-9cbe-9adc66d7fd48.png)
I believe this can be solved by scaling but I'd like to understand better what's going on before implementing anything.

<details>

```python
import numpy as np
import matplotlib.pyplot as plt

from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import SplineTransformer


x_plot = np.linspace(-1, 8, 901)
X_plot = x_plot[:, np.newaxis]

fig, ax = plt.subplots(nrows=2)
knots = [[0], [0.5], [1], [1.5], [2], [4], [4.5]]
splt_1 = SplineTransformer(
  knots=knots,
  degree=3,
  #extrapolation="periodic",
  extrapolation="constant",
  include_bias=True,
).fit(X_plot)
splt_2 = SplineTransformer(
  knots=knots,
  degree=3,
  extrapolation="periodic",
  # extrapolation="constant",
  include_bias=True,
).fit(X_plot)
ax[0].plot(x_plot, splt_1.transform(X_plot))
ax[0].legend(ax[0].lines, [f"spline {n}" for n in range(10)])
ax[0].vlines(knots, ymin=0, ymax=0.8, linestyles='dashed')
ax[1].plot(x_plot, splt_2.transform(X_plot))
ax[1].legend(ax[1].lines, [f"spline {n}" for n in range(7)])
plt.show()
```
</details>

mlondschien committed:
Add test.

mlondschien committed:
Add to changelog and example.

mlondschien committed:
Merge branch 'splines-periodic' of github.com:mlondschien/scikit-learn into splines-periodic

ogrisel commented:
Maybe you could check the continuity of the derivatives at the first and last knots explicitly in the test.

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -280,6 +293,12 @@ def fit(self, X, y=None):
+
+        # For periodic splines the spacing of the first / last degree knots
+        # needs to be equal
+        if self.extrapolation == "periodic":
+            dist_min = dist_max = (dist_min + dist_max) / 2
```
Maybe, this is the reason that the continuity does fail.
The knots "|" for non-periodic spline of degree 2 looks like
|a|b|1|2|3|4|c|d|
1-4 is the base interval, abcd the intervals for the knots beyond the base inverval.
For a periodic spline this should look like
|3|4|1|2|3|4|
The last 2 splines are removed. The first 2 knots must match the end of the period=|3|4|. |3| and |4| not only means the same knot distance, but also the same coefficient, i.e. the same spline.

What do you think?

mlondschien replied:
This did the trick. Thanks!

lorentzenchr commented:
I merged #19577. It is not strictly necessary, but it might help to merge main.

mlondschien committed:
Repeat base_knot distances. (Thanks @lorentzenchr).

mlondschien committed:
Always use floats as knots.

mlondschien committed:
Add test for smoothness.

mlondschien committed:
Add suggestions by @ogrisel.

mlondschien committed:
Merge branch 'main' into splines-periodic

mlondschien committed:
Expand explanation of periodic splines in docstring.

mlondschien committed:
Some fixes and linting.

mlondschien commented:
Adjusting the spacing to match that of the base knots did the trick. Thanks @lorentzenchr.
![Figure_1](https://user-images.githubusercontent.com/61679398/109397522-41db8600-7937-11eb-949d-b10898058763.png)

I've also added a test that manually computes derivatives and checks for smoothness. I am not to happy with this as this depends quite strongly on the tolerance.

Open points:
 - [If we increase the dependency of `scipy` to 1.0.0, we can drop our manual periodic extrapolation.](https://github.com/scikit-learn/scikit-learn/pull/19483#discussion_r583782255)
 - [This statement](https://github.com/scikit-learn/scikit-learn/pull/19483/files#diff-690d1b546a74ed69a4dc3e77edaa88363fc0f8297a71b1a3619a0f1cf79b2621R25-R27) is incorrect for periodic splines. Is this worth an extra sentence?

mlondschien committed:
Cleanup.

mlondschien committed:
Avoid scipy ValueError for integer knots.

ogrisel commented:
> This statement is incorrect for periodic splines. Is this worth an extra sentence?

Which statement? The link does not seem to work as intended. I get the generic full diff of the PR.

ogrisel commented on doc/whats_new/v1.0.rst:
```diff
@@ -192,6 +192,9 @@ Changelog
   polynomial ``degree`` of the splines, number of knots ``n_knots`` and knot
   positioning strategy ``knots``.
   :pr:`18368` by :user:`Christian Lorentzen <lorentzenchr>`.
+- |Enhancement| The :class:`preprocessing.SplineTransformer` supports periodic
+  splines via the ``extrapolation`` argument.
+  :pr:`19483` by :user:`Malte Londschien <mlondschien>`.
```
Since `SplineTransformer` is introduced in the same release, marking this as enhancement is not really correct. I would just collapse this into the previous entry.

```suggestion
  :pr:`18368` by :user:`Christian Lorentzen <lorentzenchr>`.
  :class:`preprocessing.SplineTransformer` also supports periodic
  splines via the ``extrapolation`` argument.
  :pr:`19483` by :user:`Malte Londschien <mlondschien>`.
```

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+    Xt = transformer.fit_transform(X)
+
+    tol = 0.02
+
```
Maybe add an inline comment to state that we expect the first `degree` derivatives to be continuous but `degree + 1`th derivative to be discontinuous.

mlondschien commented:
```
Generate a new feature matrix consisting of `n_splines=n_knots + degree - 1` spline basis functions (B-splines) of polynomial order=`degree` for each feature.
```
Second sentence of the docstring of the `SplineTransformer`. For periodic splines, we get `n_knots - 1` functions.

ogrisel commented:
+1 for making the docstring more specific for the periodic case.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -250,55 +253,86 @@ def fit(self, X, y=None):
+                base_knots[-(degree + 1): -1] - period,
+                base_knots,
+                base_knots[1: (degree + 1)] + period
+            ].astype(np.float)
```
This is causing the test failure. What is the dtype of `base_knots` at this point? It should already be `X.dtype` which should be fine, no? If you really need double precision, we can use `.astype(np.float64)` here instead but we should explain why. Usually scikit-learn models tends to operate at the precision level of the input data.

mlondschien replied:
If we pass knots as a list of lists of integers or an integer array, scipy complains:
```python
In [1]: from scipy.interpolate import BSpline
   ...: import numpy as np
   ...: 
   ...: BSpline.construct_fast(
   ...:     np.array([0, 1, 2, 3, 4, 5]),
   ...:     np.array([1.0, 0.0, 0.0]),
   ...:     2
   ...: )(np.array([1.0]))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-6975725b9102> in <module>
      2 import numpy as np
      3 
----> 4 BSpline.construct_fast(
      5     np.array([0, 1, 2, 3, 4, 5]),
      6     np.array([1.0, 0.0, 0.0]),

~/anaconda3/envs/sklearn/lib/python3.8/site-packages/scipy/interpolate/_bsplines.py in __call__(self, x, nu, extrapolate)
    336         out = np.empty((len(x), prod(self.c.shape[1:])), dtype=self.c.dtype)
    337         self._ensure_c_contiguous()
--> 338         self._evaluate(x, nu, extrapolate, out)
    339         out = out.reshape(x_shape + self.c.shape[1:])
    340         if self.axis != 0:

~/anaconda3/envs/sklearn/lib/python3.8/site-packages/scipy/interpolate/_bsplines.py in _evaluate(self, xp, nu, extrapolate, out)
    346 
    347     def _evaluate(self, xp, nu, extrapolate, out):
--> 348         _bspl.evaluate_spline(self.t, self.c.reshape(self.c.shape[0], -1),
    349                 self.k, xp, nu, extrapolate, out)
    350 

_bspl.pyx in scipy.interpolate._bspl.evaluate_spline()

ValueError: Buffer dtype mismatch, expected 'const double' but got 'long'
```

This hasn't been an issue before as `np.linspace` returns `floats` and `knots` get casted. [This test](https://github.com/scikit-learn/scikit-learn/pull/19483/commits/b8da79ff21e20fa029256b0a744bf57712f68404#diff-5527bc676d1f3e3d77d2964f863103039ef46f809db3a41ffaea58d5d686fb1bR90-R107) fails for periodic extrapolation. We could avoid the `.astype(np.float)` by using the default constructor of `BSpline`.

lorentzenchr replied:
`base_knots` are already `np.float64` as returned by `_get_base_knot_positions`. The reason is, again, that `BSpline` converts them anyway:
https://github.com/scipy/scipy/blob/5ab7426247900db9de856e790b8bea1bd71aec49/scipy/interpolate/_bsplines.py#L176

I think, the explicit conversion here is unnecessary.

mlondschien replied:
Not if the knots get passed explicitly (instead of being constructed with `_get_base_knot_positions`).

lorentzenchr replied:
Right. I passed manually, `check_array` only makes sure that `base_knots` is numeric.
I think, the simplest would be to pass `dtype=np.float64` to `check_array`.

ogrisel commented:
@mlondschien I would appreciate it if you could explain how to use this correctly in practice.  Do you have a public ML benchmark dataset (e.g. on openml.org) where this extrapolation capabilities can naturally give an edge over more traditional feature engineering?

Intuitively, to model periodic effects in a predictive model, I would rather just apply the traditional `SplineTransformer` on features that are naturally periodic and therefore with no need for extrapolation, for instance on features that represent "day of year" and "day of week".

I am failing to see where the empirical range of non-periodic feature observed on the training set could serve successfully as the implicit period for extrapolation.

ogrisel commented:
Actually, now I see it. Even if you use naturally periodic features such as "day or year" or "day of week", using the periodic extrapolation might still be beneficial even if those feature value never go out of range because of the smoothness of the derivative of the first and last knot to avoid a jump in the feature value between the end of a year (Dec 31) and the beginning of the next year (Jan 1).

It might be worth explaining this with an additional paragraph in the example or the user guide.

mlondschien committed:
Update doc/whats_new/v1.0.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -116,7 +120,6 @@ class SplineTransformer(TransformerMixin, BaseEstimator):
            [0.  , 0.08, 0.74, 0.18],
            [0.  , 0.  , 0.5 , 0.5 ]])
     """
-
```
Please revert, see https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings.

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -250,55 +253,86 @@ def fit(self, X, y=None):
+
+        if self.extrapolation == "periodic":
+            # For periodic splines the spacing of the first / last degree knots
+            # needs to a continuation of the base_knot spacing.
```
Can you make this a bit more precise, in particular "needs to a continuation"?

mlondschien committed:
np.float64.

mlondschien committed:
More specific docstring.

mlondschien committed:
Add paragraph about naturally periodic features to example.

lorentzenchr commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -250,55 +253,86 @@ def fit(self, X, y=None):
-                num=degree,
-            ),
-        ]
+
```
```suggestion
```

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -74,14 +87,24 @@ def test_spline_transformer_input_validation(params, err_msg):
         SplineTransformer(**params).fit(X)
 
 
-def test_spline_transformer_manual_knot_input():
-    """Test that array-like knot positions in SplineTransformer are accepted.
+@pytest.mark.parametrize("extrapolation", ["continue", "periodic"])
+def test_spline_transformer_manual_knot_input(extrapolation):
+    """
+    Test that array-like knot positions in SplineTransformer are accepted.
     """
     X = np.arange(20).reshape(10, 2)
-    knots = [[0.5, 1], [1.5, 2], [5, 10]]
-    st1 = SplineTransformer(degree=3, knots=knots).fit(X)
+    knots = [[0, 1], [1, 2], [5, 10], [6, 11]]
```
Why the change of knot positions?

mlondschien replied:
See this discussion:
https://github.com/scikit-learn/scikit-learn/pull/19483/files/4aaf4414c89b04e9f70fe09cdf576ac5c7a9c79e#r584184962

scipy was throwing an error when using `construct_fast` and integer value knot positions. I've changed the knots positions to cover this case.

lorentzenchr replied:
But before, there were floats like `0.5` and `1.5`, now its integers only. On top, it should work with regardless of dtype of knots, as long as numeric.

mlondschien replied:
I was lazy and wanted to cover both cases (list and integer values) with the same test. I've added a separate test and reverted the knots (the GitHub UI is displaying this incorrectly).

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+def test_spline_transformer_periodic_spline_backport():
+    """Test that the backport of extrapolate="periodic" works correctly"""
+    X = np.linspace(-2, 3.5, 10)[:, None]
+
```
```suggestion
    degree = 2

```

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+
+    # Use periodic extrapolation backport in SplineTransformer
+    transformer = SplineTransformer(
+        degree=2,
```
```suggestion
        degree=degree,
```

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+
+    # Use periodic extrapolation in BSpline
+    coef = np.array([[1.0, 0.0], [0.0, 1.0], [1.0, 0.0], [0.0, 1.0]])
+    spl = BSpline(np.arange(-3, 4), coef, 2, "periodic")
```
```suggestion
    spl = BSpline(np.arange(-3, 4), coef, degree, "periodic")
```

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+
+    tol = 0.02
+
+    for d in range(1, degree + 1):
+        Xt = Xt[1:, :] - Xt[:-1, :]
+        assert (np.abs(Xt) < (tol ** d)).all()
```
```suggestion
    for d in range(1, degree + 1):
        dXt = Xt[1:, :] - Xt[:-1, :]
        assert (np.abs(dXt) < (tol ** d)).all()
```
Why the loop with `d` and not just `d = degree`?

mlondschien replied:
See the (new) inline comments. The first `degree` numeric derivatives should be (relatively) small, and the `degree + 1`-th derivative big.

lorentzenchr replied:
Now, I get it. As the stepsize of X is `(10 - 0) / 1000 = 0.01`, shouldn't `tol = 0.01`?

mlondschien replied:
I did it properly now.

mlondschien commented on examples/linear_model/plot_polynomial_interpolation.py:
```diff
@@ -145,3 +146,63 @@ def f(x):
+# extrapolation beyond the range of the training observations. In some
+# settings, e.g. with seasonal effects, we expect a periodic continuation of
+# the underlying signal. Such effects can be modelled using periodic splines,
+# which have equal function value and equal derivatives at the first and last
```
I put _derivatives_ as the first `degree - 1` derivatives match.

mlondschien commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +191,105 @@ def test_spline_transformer_linear_regression(bias, intercept):
+        Xt = Xt[1:, :] - Xt[:-1, :]
+        assert (np.abs(Xt) < (tol ** d)).all()
+
+    assert np.abs(Xt[1:, :] - Xt[:-1, :]).max() > tol ** (degree + 1)
```
```suggestion
    assert np.abs(dXt[1:, :] - dXt[:-1, :]).max() > tol ** (degree + 1)
```

mlondschien committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

mlondschien committed:
Implement suggestions by @lorentzenchr.

mlondschien committed:
Merge branch 'splines-periodic' of github.com:mlondschien/scikit-learn into splines-periodic

mlondschien committed:
dXt.

mlondschien committed:
Revert and add test.

mlondschien committed:
Revert dtype casting to np.float64.

mlondschien committed:
Linting.

mlondschien committed:
Simplify test.

mlondschien committed:
Add float64 dtype casting back in.

mlondschien commented:
> I am not to happy with this as this depends quite strongly on the tolerance.

Here is a plot of the first 4 numerical derivatives of a degree 3 (periodic) spline:
![Figure_1](https://user-images.githubusercontent.com/61679398/109416958-bbb85180-79c1-11eb-8f26-2c57936c592a.png)

The test works quite well with the current tolerance.

>  Do you have a public ML benchmark dataset (e.g. on openml.org) where this extrapolation capabilities can naturally give an edge over more traditional feature engineering?

No. But your other comment explains the benefits quite well. Typically the main benefit of using periodic splines is not prediction performance (which would be rather minor in most cases), but rather mathematical correctness (e.g. the model does no longer depend on whether the week starts on Sunday or Monday) and explainability. I've added a paragraph to the example.

> +1 for making the docstring more specific for the periodic case.

I've added a sentence.

mlondschien committed:
Cover case with multiple features.

mlondschien commented:
I don't fully understand the use-case of multiple input features, but I guess this is standard for scikit-learn transformers? I've added the integer input knots case as a separate test and reverted the `test_spline_transformer_manual_knot_input`. What's up with the failing CI? Do jobs get cancelled upon a new commit? I tried to highlight the failing test after removing the `astype(np.float64)` (without success it seems).

mlondschien committed:
Different tolerance for smoothness property.

mlondschien committed:
Pass np.float64 as dtype argument to check_array.

mlondschien committed:
Fix test.

ogrisel commented:
> Do jobs get cancelled upon a new commit? 

Yes to spare compute resources.

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +194,109 @@ def test_spline_transformer_linear_regression(bias, intercept):
+        dXt = dXt[1:, :] - dXt[:-1, :]
+        assert (np.abs(dXt) < tol * (delta ** d)).all()
+
+    assert np.abs(dXt[1:, :] - dXt[:-1, :]).max() > tol * delta ** (degree + 1)
```
```suggestion
    # The (degree + 1)-th order derivative is not continuous.
    dXt =  dXt[1:, :] - dXt[:-1, :]
    assert np.abs().max(dXt) > tol * delta ** (degree + 1)
```

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +194,109 @@ def test_spline_transformer_linear_regression(bias, intercept):
+@pytest.mark.parametrize("degree", [3, 5])
+def test_spline_transformer_periodic_splines_smoothness(degree):
+    """Test that spline transformation is smooth at first / last knot."""
+    X = np.linspace(0, 10, 1000)[:, None]
```
Maybe you could start the range before the first knot to actually check the smoothness at 0.0:

```suggestion
    X = np.linspace(-2, 10, 1000)[:, None]
```

ogrisel commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +194,109 @@ def test_spline_transformer_linear_regression(bias, intercept):
+        dXt = dXt[1:, :] - dXt[:-1, :]
+        assert (np.abs(dXt) < tol * (delta ** d)).all()
+
+    assert np.abs(dXt[1:, :] - dXt[:-1, :]).max() > tol * delta ** (degree + 1)
```
For consistency with the previous line.

```suggestion
    # The (degree + 1)-th order derivative is not continuous.
    dXt =  dXt[1:, :] - dXt[:-1, :]
    assert np.abs(dXt).max() > tol * delta ** (degree + 1)
```

mlondschien replied:
I've added a bunch of comments.

ogrisel commented on examples/linear_model/plot_polynomial_interpolation.py:
```diff
@@ -145,3 +146,67 @@ def f(x):
+# Periodic splines can also be useful for naturally periodic features (such as
+# day of the year), as the smoothness at the boundary knots prevents a jump in
+# the transformed values (e.g. from Dec 31st to Jan 1st).
+
```
Maybe add a sentence such as: For such naturally periodic features, it is advised to manually set the knots values to make sure that the periodicity is not by a random subsampling effects of the training set, for instance in a cross-validation loop.

ogrisel reviewed (approved):
Thanks for your feedback. I now better see the value of this option. Here are some more suggestions.

Even if I do not follow the maths, the tests, the plots and the example look convincing to me.

ogrisel commented:
> I don't fully understand the use-case of multiple input features, but I guess this is standard for scikit-learn transformers? 

scikit-learn transformers generally process a group of features at once, for instance PCA. Transformers that treats one input feature at a time have to follow the same API for consistency. Hence the input data is always shaped `(n_samples, n_features)`.

ogrisel commented:
Actually it might be worth adding a test that check the periodicity extrapolation of the transformer (without  sticking it into a pipeline with a regressor) with 2 input features e.g. feature 0 with values in [-1, 1] and features 1 with values in [0, 5] in the training set.

And then checking that we get the transformed features on a test set with values on [1, 2] for features 0 and [5, 10] for feature 1.

lorentzenchr commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -168,6 +194,109 @@ def test_spline_transformer_linear_regression(bias, intercept):
+    Xt = transformer.fit_transform(X)
+
+    delta = (X.max() - X.min()) / len(X)
+    tol = 10
+
+    dXt = Xt
+
+    for d in range(1, degree + 1):
+        dXt = dXt[1:, :] - dXt[:-1, :]
+        assert (np.abs(dXt) < tol * (delta ** d)).all()
```
```suggestion
    tol = 1e-3
    
    # continuity of function
    assert (np.abs(np.diff(Xt, axis=0)) < tol).all()
    
    # continuity of derivatives
    dXt = Xt
    for d in range(1, degree + 1):
        dXt = np.diff(dXt, axis=0) / delta  # d-th derivative
        # continuity of d-th derivative
        assert (np.abs(np.diff(dXt, axis=0)) < tol).all()
```
I hope this works. This is just a proposal to make this test cleaner.

mlondschien replied:
I'm now using `np.diff` and added a bunch of comments. Note that your proposal checks for the continuity of `Xt` twice, once on L292 and once in L299 for d = 1. Also `tol` needs to be of the order of (or rather bigger than) the expected size of derivatives. See the plot above. Since we're seeing values around 1, I've set `tol = 10`.

ogrisel replied:
I addressed some of this in https://github.com/scikit-learn/scikit-learn/pull/19483#issuecomment-788807414 below.

lorentzenchr commented:
Periodic b-splines turn out to be surprisingly complicated. But we're close to the finish.

mlondschien committed:
Use np.diff, test at zero and add comments to smoothness test.

mlondschien committed:
Expand example text.

mlondschien committed:
Add test.

mlondschien commented:
> Yes to spare compute resources.

This makes sense. I would have expected the tests to never finish (i.e. yellow dot) instead of failing though.

> Periodic b-splines turn out to be surprisingly complicated. But we're close to the finish.

Thanks for investing the time to review.

mlondschien committed:
Use sklearn.fixes.linspace.

ogrisel committed:
Improve the spline smoothness test

ogrisel commented:
I pushed ca92728 to improve `test_spline_transformer_periodic_splines_smoothness` to test continuity on the raw diff, prior to dividing by delta (to get the derivative). Also I changed the assertion by comparing the scaler abs max to tol to get more informative values in the pytest assertion error message in case of failure.

mlondschien commented on sklearn/preprocessing/tests/test_polynomial.py:
```diff
@@ -322,18 +322,21 @@ def test_spline_transformer_periodic_splines_smoothness(degree):
         # Check continuity of the (d-1)-th derivative
-        assert (np.abs(dXt) < tol).all()
+        diff = np.diff(dXt, axis=0)
+        assert np.abs(diff).max() < tol
```
Note that this (and below) depends on `tol / delta ~ 10`. I.e. if we reduce (or increase) the number of rows in `X` substantially, this test will fail.

ogrisel replied:
I don't understand why though. Continuity should always be checked on the raw diff, assuming that the steps are fine grained enough, no?

Otherwise we "check continuity" by bounding the absolute value of the derivative itself which seems wrong to me.

mlondschien replied:
Given continuity, the "raw diff" scales with the order of the stepsize (`delta`). It will be roughly of the same order. Above, your stepsize is ~1e-3, so checking against `tol=1e-2` is fine. If we reduce the stepsize to 0.1, the diff will be of order `diff ~ 0.1 > tol  = 1e-2`.

mlondschien commented:
@lorentzenchr, if you could have another (final 😉?) look, that would be great.

lorentzenchr commented:
@mlondschien This PR looks very good and, if necessary, we can improve some minor things in later PRs. I have 2 remaining issues, a small one:
- Add the periodic case to `test_spline_transformer_unity_decomposition`.

And a large one: `test_spline_transformer_periodic_splines_smoothness`
`tol=1e-2` is not a very strict test. Also, we don't need to test continuity at all values of X. Just at the boundaries `b` of the period. Something like
```python
eps = 1e-8
for b in [0, 8, 16]:
    Xnew = np.linspace(b - degree*eps , b + degree*eps, num=2*degree + 1)[:, None]
    Xt = transformer.fit_transform(Xnew).squeeze()
    assert Xt[degree-1]  = pytest.approx(Xt[degree+1])  # compare at b-eps and b+eps
    # TODO: compare left n-th derivative at b-eps with right n-th derivative at b+eps
```
What do you think?

mlondschien commented:
> * Add the periodic case to `test_spline_transformer_unity_decomposition`.

Sure!

> `tol=1e-2` is not a very strict test. Also, we don't need to test continuity at all values of X. Just at the boundaries `b` of the period. Something like
> 
> ```python
> eps = 1e-8
> for b in [0, 8, 16]:
>     Xnew = np.linspace(b - degree*eps , b + degree*eps, num=2*degree + 1)[:, None]
>     Xt = transformer.fit_transform(Xnew).squeeze()
>     assert Xt[degree-1]  = pytest.approx(Xt[degree+1])  # compare at b-eps and b+eps
>     # TODO: compare left n-th derivative at b-eps with right n-th derivative at b+eps
> ```
> 
> What do you think?

IMO the test from before, I.e. with `X.shape = (1000,1)`, was strict enough, as we checked that the diff was an order of magnitude above the tolerance for the last iteration where we expected discontinuity. This test makes sure that we did not choose a too large tolerance. However, if you prefer, I can also adjust the test to only check continuity at the boundary, choose a smaller stepsize there and thus a smaller tolerance. We could then also test continuity of the constant extrapolation and C1 of linear extrapolation in the same test via parametrization.

mlondschien committed:
Add periodic splines to unity decomposition test.

mlondschien commented:
I have adjusted (but not pushed) the smoothness test as proposed by you @lorentzenchr. See here:
```python
@pytest.mark.parametrize("degree", [3, 5])
def test_spline_transformer_periodic_splines_smoothness(degree):
    """Test that spline transformation is smooth at first / last knot."""
    X = np.linspace(0, 8, 10)[:, None]

    transformer = SplineTransformer(
        degree=degree,
        extrapolation="periodic",
        knots=[[0.0], [1.0], [3.0], [4.0], [5.0], [8.0]]
    )
    Xt = transformer.fit(X)

    eps = 1e-8

    for b in [0, 8, 16]:
        dXt = transformer.transform(
            np.linspace(
                b - degree * eps,
                b + degree * eps,
                2 * degree + 1
            )[:, None]
        )

        for d in range(degree):
            assert_allclose(dXt[0, :], dXt[-1, :], atol = 10 * eps, rtol=0)
            dXt = np.diff(dXt, axis=0) / eps
```
This fails due to problems with precision. The original `dXt` has precision of around `1e-16` and so has `np.diff(dXt, axis=0)`. In the second iteration, due to dividing by `eps`, the `dXt` has precision of around `1e-8`. As you can imagine, the test fails for `degree >= 3`.

Possible options:
 - Use lower precision here, such that `eps ** degree >> 1e-16`. For `degree = 5`, this implies `eps > 1e-3`.
 - Go back to the original test.

Let me know what you think @lorentzenchr. If you had something different in mind that does not have this problem, please let me know. I don't believe that we should hold off on merging this PR due to details about tolerances / precision. Feel free to add any changes you would like to the smoothness test onto this PR.

mlondschien committed:
flake8.

lorentzenchr commented:
What about looking at `b-eps` and `b+eps` with `eps=1e-12`, but then use a different variable, i.e. `h=1e-3`, for the numerical drivatives. I'm really only interested in the 2 values at `b-eps` and `b+eps`. In addition, one can use higher order formulas for numerical derivatives, see https://en.wikipedia.org/wiki/Finite_difference_coefficient and https://github.com/lorentzenchr/scikit-learn/blob/4e007384e7eb4e86553a1bb44a220d9a7ce48134/sklearn/_loss/tests/test_loss.py#L89-L104.
Ideally, one would use backward differences at `b-eps` and forward differences at `b+eps`.

ogrisel commented on sklearn/preprocessing/_polynomial.py:
```diff
@@ -54,14 +54,17 @@ class SplineTransformer(TransformerMixin, BaseEstimator):
+        `extrapolate=True` in :class:`scipy.interpolate.BSpline`. If
+        'periodic', periodic splines with a periodicity equal to the distance
+        between the first and last knot are used. Periodic splines enforce
+        equal function values and derivatives at the first and last knot.
```
I think we could be extra helpful and explain why it's a nice property on a practical use case. For instance:

```suggestion
        equal function values and derivatives at the first and last knot.
        For example, this makes it possible to avoid introducing an arbitrary
        jump between Dec 31st and Jan 1st in splines features derived from a
        naturally periodic "day-of-year" input feature. In this case it is
        recommend to manually set the knot values to control the period.
```

Maybe it's a bit verbose for a docstring but it took quite a long time for me to get it, I think it's better to be extra explicit and many of our users might find it useful.

Alternatively we could have a similar paragraph in the user guide (under `doc/modules/preprocessing.rst`).

mlondschien replied:
I don't mind verbosity at all.

mlondschien commented:
> In addition, one can use higher order formulas for numerical derivatives

I could not find an implementation for forward / backward diff weights in scipy. Are you fine with [`scipy.misc.central_diff_weights`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.central_diff_weights.html)?

mlondschien committed:
Update sklearn/preprocessing/_polynomial.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

lorentzenchr commented:
> I could not find an implementation for forward / backward diff weights in scipy. Are you fine with [`scipy.misc.central_diff_weights`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.central_diff_weights.html)?

It's not my intention to make this PR more complex. If you could use https://en.wikipedia.org/wiki/Finite_difference_coefficient, that'd be great. Using `scipy.misc.central_diff_weights`might also do the job. So I'd suggest use that and make a comment such that a future PR could improve this with forward and backward differences.

mlondschien committed:
Make tol dependent of delta.

mlondschien commented:
> It's not my intention to make this PR more complex.

To be frank, IMO its too late for this. I have probably spent as much time on the smoothness test as on the periodic splines themselves.

>  If you could use https://en.wikipedia.org/wiki/Finite_difference_coefficient, that'd be great. Using `scipy.misc.central_diff_weights` might also do the job. So I'd suggest use that and make a comment such that a future PR could improve this with forward and backward differences.

How would you go about using the coefficients from Wikipedia? Just copy them over?

I have tried using (central) finite differences as suggested by you. See the code in the details below. The test still fails for high degrees / small tolerances. IMO this is still due to precision. According to Wikipedia, the `d`-th derivative can be approximated as the sum of products of weights and function evaluation divided by the stepsize (`eps`) to the `d`-th power. This is also how `scipy.misc.derivative` is implemented. According to Wikipedia, if I use order (accuracy) `o` coefficients, the approximation will be correct up to a term of order stepsize to the `o`-th power. However, IIUC, this assumes infinite precision in the function evaluations themselves. In our use-case, the function evaluations have precision around `1e-16`, so using a stepsize of `eps`, any approximation of the `d`-th derivative will have a precision of at best `1e-16 * eps ** d`.

See the following example:
```
In [1]: from scipy.misc import derivative
   ...: import numpy as np
   ...: 
   ...: def f(x):
   ...:     return np.sin(x)
   ...: 

In [2]: derivative(f, 0, dx=1e-8, n=1, order=7)
Out[2]: 1.0000000000000002

In [3]: derivative(f, 0, dx=1e-8, n=2, order=7)
Out[3]: 7.237830359838991e-09

In [4]: derivative(f, 0, dx=1e-8, n=3, order=7)
Out[4]: 9.098986738083303

In [5]: derivative(f, 0, dx=1e-4, n=3, order=7)
Out[5]: -0.9999999011365072

In [6]: derivative(f, 0, dx=1e-4, n=5, order=7)
Out[6]: 5.421010862427521

In [7]: derivative(f, 0, dx=1e-2, n=5, order=7)
Out[7]: 0.999966567882815
```
Note how the third derivative of `f` using a stepsize of `1e-8` is complete garbage. Reducing the stepsize to `1e-4` produces a much more reasonable result.

I believe that hunting higher tolerances is useless. IMO the test as I had implemented it originally or as it is currently is sufficient for the use-case. I have added a small commit onto @ogrisel's suggestion to make the tolerance dependent of `delta`. I am not willing to invest a significant amount of time into the test. If you are unhappy with the tolerance, please add a commit on top to fix this. One option would be to only test until `degree=2` with stepsize `eps=1e-8` or until `degree = 4` with stepsize `eps=1e-4`. I believe that asserting the extrapolation is smooth for higher degrees is more important than a strict tolerance.

I sincerely hope that this test does not prevent us from merging the PR. Periodic splines are a feature that improves on the status quo. The current tests are much more extensive than those before, e.g. as there are no tests for continuity / differentiability at the border for constant / linear extrapolation.

<details>

```python

@pytest.mark.parametrize("degree", [3, 5])
def test_spline_transformer_periodic_splines_smoothness(degree):
    """Test that spline transformation is smooth at first / last knot."""
    X_fit = np.linspace(-2, 10, 10)[:, None]

    transformer = SplineTransformer(
        degree=degree,
        extrapolation="periodic",
        knots=[[0.0], [1.0], [3.0], [4.0], [5.0], [8.0]]
    ).fit(X_fit)

    eps = 1e-8

    # We use scipy.misc.central_diff_weights to compute higher-order
    # derivatives of sufficient accuracy. See also
    # https://en.wikipedia.org/wiki/Finite_difference_coefficient
    weights = {
        d: central_diff_weights(2 * d + 1, d) for d in range(1, degree + 1)
    }

    for b in [0, 8, 16]:
        X_left = np.linspace(
            b - (2 * degree + 1) * eps,
            b - eps,
            2 * degree + 1
        )[:, None]
        X_right = np.linspace(
            b + eps,
            b + (2 * degree + 1) * eps,
            2 * degree + 1
        )[:, None]
        Xt_left = transformer.transform(X_left)
        Xt_right = transformer.transform(X_right)

        for d in range(1, degree + 1):
            dXt_left = np.matmul(
                Xt_left[(degree - d) : (degree + d + 1), :].T,
                weights[d]
            ) / (eps ** d)
            dXt_right = np.matmul(
                Xt_right[(degree - d) : (degree + d + 1), :].T,
                weights[d]
            ) / (eps ** d)
            assert_allclose(dXt_left, dXt_right, rtol = 0, atol=10)
```

ogrisel commented:
I agree let's merge. Exploring how to best test derivatives in scikit-learn in general can be done in another PR.

lorentzenchr reviewed (approved):
LGTM. @mlondschien Thank you so much. This is a very useful extension. And sorry for being such a p.i.t.a. about the continuity test.

lorentzenchr commented:
@mlondschien Also thanks for being frank and patience.

ogrisel commented:
Thank you very much for the very nice contribution @mlondschien!
